src/Polysemy/Final/Internal.hs

```diff
@@ -0,0 +1,211 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Polysemy.Final.Internal where
+
+import Data.Functor.Compose
+import Data.Maybe
+
+import Control.Concurrent.MVar
+import Control.Exception
+import Control.Monad
+import Control.Monad.Trans.Maybe
+
+import Polysemy
+import Polysemy.Internal
+import Polysemy.Internal.Union
+
+-----------------------------------------------------------------------------
+-- | An effect for embedding higher-order effects in the final target monad
+-- of the effect stack.
+--
+-- This is very useful for writing interpreters that interpret higher-order
+-- effects in terms of the final monad.
+--
+-- 'Final' is slightly weaker, but using it is otherwise more preferable
+-- than using a provided lowering function (a la 'Polysemy.Async.lowerAsync'):
+--
+--   * Interpreters using 'Final' may be run and composed like any other
+--     interpreter, unlike @lower-@ interpreters which need to be
+--     composed using '.@' or '.@@'.
+--
+--   * Multiple return-value changing interpreters using 'Final' may be
+--     composed together without issue. It's very difficult to
+--     compose 'Polysemy.Error.lowerError' with itself, unlike
+--     'Polysemy.Final.Error.errorToIOFinal'.
+--
+--   * Initialization work of @lower-@ interpreters may be duplicated
+--     when composed together with '.@'. @-'Final'@ interpreters avoid
+--     this issue altogether, as long as the initialization work is performed
+--     outside of a 'interpretH'/'interpretFinal'.
+--
+--   * Instead of having access to a natural transformation
+--     @forall x. 'Sem' r x -> m x@, @'Final' m@ provides a
+--     distribution through the effectful state:
+--     @forall x. f ('Sem' r x) -> m (f x)@,
+--     together with the initial state @f ()@, and an @'Inspector' f@.
+--     This is powerful enough for most purposes.
+--
+-- /Beware/: 'Final' actions are interpreted as actions of the final monad,
+-- and the effectful state visible to
+-- 'withWeaving'\/'Polysemy.Final.withStrategic'\/'Polysemy.Final.interpretFinal'
+-- is that of /all interpreters run in order to produce the final monad/.
+--
+-- This means that any interpreter built using 'Final' will /not/
+-- respect local/global state semantics based on the order of
+-- interpreters run. You should signal interpreters that make use of
+-- 'Final' by adding a @-'Final'@ suffix to the names of these.
+--
+-- State semantics of effects that are /not/
+-- interpreted in terms of the final monad will always
+-- appear local to effects that are interpreted in terms of the final monad.
+--
+-- State semantics between effects that are interpreted in terms of the final monad
+-- depend on the final monad. For example, if the final monad is a monad transformer
+-- stack, then state semantics will depend on the order monad transformers are stacked.
+newtype Final m z a where
+  WithWeaving :: (  forall f
```

Comment:
How about wrapping this scary thing in a GADT so we can name all of its parameters?

Reply:
You mean so it'd become something like this?:
```haskell
newtype Final m z a where
  WithWeaving :: (forall f. Functor f => WithWeavingState f m -> m (f a)) 
                        -> WithWeaving m a
```
Where `WithWeavingState` is a record?

If the goal is to improve documentation, I'd instead prefer this, if it works with Haddock (I have issues trying to run Haddock, so I can't check):

```haskell
newtype Final m z a where
  WithWeaving :: (  forall f
                  . Functor f
                 => f ()
                    -- ^ Initial effectful state. Blablabla...
                 -> (forall x. f (z x) -> m (f x))
                    -- ^ Conversion with effectful state to the final monad. Blablabla...
                 -> (forall x. f x -> Maybe x)
                    -- ^ Inspector. Blablabla...
                 -> m (f a)
                 )
              -> Final m z a
```

Reply:
The goal is for documentation. Unfortunately haddock is too dumb to annotate higher rank functions; but I'd be happy with comments here if you feel the GADT is too heavyweight.

Reply:
I mean, I kinda feel that way? I like the mantra of `withWeaving $ \s wv ins -> ...` once you understand how it works.

Besides, even though we won't be able to annotate the constructor, we will be able to annotate the derived `withWeaving` action.

By the way, something that just cropped up in my mind; should `WithWeaving` be renamed to `WithWeavingToFinal`? More verbose, but also mirrors `withLowerToIO`.

Reply:
Both sound good to me.